yt_dlp/extractor/tubetugraz.py

```diff
@@ -0,0 +1,349 @@
+            episode_info,
+            lambda x: x["mediapackage"]["media"]["track"]) or []
+
+        formats = []
+        format_types = defaultdict(lambda: defaultdict(int))
+        for format_info in format_infos:
+            formats.extend(self._extract_formats(format_info, format_types))
+
+        self._guess_formats(formats, format_types, id)
+        self._sort_formats(formats)
+
+        return {
+            "_type": "video",
+            "id": id,
+            "title": title,
+            "creator": creator,
+            "duration": duration,
+            "series": series_title,
+            "episode": episode_title,
+            "formats": formats
+        }
+
+    def _extract_formats(self, format_info, format_types):
+        PREFERRED_TYPE = "presentation"
+
+        url = try_get(format_info, lambda x: x["url"])
+        type = try_get(format_info, lambda x: x["type"]) or "unknown"
+        transport = try_get(format_info, lambda x: x["transport"]) or "https"
+        audio_bitrate = try_get(format_info, lambda x: x["audio"]["bitrate"])
+        video_bitrate = try_get(format_info, lambda x: x["video"]["bitrate"])
+        framerate = try_get(format_info, lambda x: x["video"]["framerate"])
+        resolution = try_get(format_info, lambda x: x["video"]["resolution"])
+
+        type = str(type).replace("/delivery", "")
+        transport = str(transport).lower()
+
+        if isinstance(audio_bitrate, int):
+            audio_bitrate = audio_bitrate / 1000
+        if isinstance(video_bitrate, int):
+            video_bitrate = video_bitrate / 1000
+        if isinstance(audio_bitrate, int) and isinstance(video_bitrate, int):
+            bitrate = audio_bitrate + video_bitrate
+        else:
+            bitrate = None
+
+        if type == PREFERRED_TYPE:
+            preference = -1
+        else:
+            preference = -2
+
+        if transport == "https":
+            formats = [{
+                "url": url,
+                "tbr": bitrate,
+                "abr": audio_bitrate,
+                "vbr": video_bitrate,
+                "framerate": framerate,
+                "resolution": resolution,
+            }]
+        elif transport == "hls":
+            formats = self._extract_m3u8_formats(
+                url, None,
+                note="downloading %s HLS manifest" % type,
+                fatal=False,
+                ext="mp4")
+        elif transport == "dash":
+            formats = self._extract_mpd_formats(
+                url, None,
+                note="downloading %s DASH manifest" % type,
+                fatal=False)
+        else:
+            # RTMP, HDS, SMOOTH, and unknown formats
+            # - RTMP url fails on every tested entry until now
+            # - HDS url 404's on every tested entry until now
+            # - SMOOTH url 404's on every tested entry until now
+            formats = []
+
+        for format in formats:
+            format["preference"] = preference
```

Comment:
`preference` should generally not be set. Can you explain the diff b/w the two types of formats? From my understanding, `source_preference` makes more sense?

Reply:
tube.tugraz.at has many lectures that have both a camera stream from the lecture hall ("PRESENTER"), showing the lecturer and the blackboard, as well as a video stream from the connected computer ("PRESENTATION"), usually showing a powerpoint presentation or similar.

The majority of lecture videos uses only PRESENTATION streams, or has a PRESENTER stream that doesn't really show anything useful, and most lecturers that use the blackboard don't use a PRESENTATION stream at all.

These are two separate video streams that can be watched side-by-side on the website if both are present. As far as I understand it, it's hard to map this to YT-DLP's model, so making a choice of which stream to choose by default is necessary if consistent behaviour is desired.

Choosing the best PRESENTATION stream by default is most likely the most sane option, because it will almost always deliver a video that actually contains the content you want to watch. In cases where you want a specific stream, you can still choose.